src/github.com/matrix-org/dendrite/syncapi/sync/requestpool.go

```diff
@@ -62,48 +62,71 @@ func (rp *RequestPool) OnIncomingSyncRequest(req *http.Request, device *authtype
 		"timeout": syncReq.timeout,
 	}).Info("Incoming /sync request")
 
-	// Fork off 2 goroutines: one to do the work, and one to serve as a timeout.
-	// Whichever returns first is the one we will serve back to the client.
-	timeoutChan := make(chan struct{})
-	timer := time.AfterFunc(syncReq.timeout, func() {
-		close(timeoutChan) // signal that the timeout has expired
-	})
-
-	done := make(chan util.JSONResponse)
-	go func() {
-		currentPos := rp.notifier.WaitForEvents(*syncReq)
-		// We stop the timer BEFORE calculating the response so the cpu work
-		// done to calculate the response is not timed. This stops us from
-		// doing lots of work then timing out and sending back an empty response.
-		timer.Stop()
-		syncData, err := rp.currentSyncForUser(*syncReq, currentPos)
-		var res util.JSONResponse
+	currPos := rp.notifier.CurrentPosition()
+
+	// If this is an initial sync or timeout=0 we return immediately
+	if syncReq.since == types.StreamPosition(0) || syncReq.timeout == 0 {
+		syncData, err := rp.currentSyncForUser(*syncReq, currPos)
 		if err != nil {
-			res = httputil.LogThenError(req, err)
-		} else {
-			syncData, err = rp.appendAccountData(syncData, device.UserID, *syncReq, currentPos)
-			if err != nil {
-				res = httputil.LogThenError(req, err)
-			} else {
-				res = util.JSONResponse{
-					Code: 200,
-					JSON: syncData,
-				}
-			}
+			return httputil.LogThenError(req, err)
 		}
-		done <- res
-		close(done)
-	}()
-
-	select {
-	case <-timeoutChan: // timeout fired
 		return util.JSONResponse{
 			Code: 200,
-			JSON: types.NewResponse(syncReq.since),
+			JSON: syncData,
 		}
-	case res := <-done: // received a response
-		return res
 	}
+
+	// Otherwise, we wait for the notifier to tell us if something *may* have
+	// happened. We loop in case it turns out that nothing did happen.
+
+	timeoutChan := time.After(syncReq.timeout) // case of timeout=0 is handled above
```

Comment:
is it safe to return without reading the time value from this? There isn't going to be some memory leak or something?

Reply:
Sigh, I thought the GC would do something useful, since pretty much everywhere seems to suggest doing timeouts like this. But I've found how to do it with explicit clean up.

Reply:
"everywhere suggests doing it like X" doesn't mean that X is a sane solution, in my bitter experience :/

Reply:
Yeah, but when its gobyexample.com, golang blog and the golang wiki, I was inclined to trust them :/